### ISSUE TITLE
Noooooo my .git is dead! / Ignore dotfiles?

### ISSUE THREAD

**Issue by testerez** [prettier/prettier#2882]
I was just trying prettier to see the result on my codebase. I should have been more cautious but I scrapped my `.git` directory. All my local branches are probably lost forever!

You should consider adding a check to prevent altering any file with `.git` in the path.

😢

**Comment by testerez** [comment]
Actually, it seems like it was just my `.git/HEAD` that was scrapped. Fixed it and I can see my branches again 😅 
Still I think that securing `.git` directory would be a good idea.

**Comment by lydell** [comment]
I don't understand what happened, and what you're proposing that Prettier should do?

**Comment by testerez** [comment]
Ok, here is how to reproduce my issue:

```
mkdir test
cd test
git init
npx prettier --write "**/*"
git status
```
Result: `fatal: Not a git repository (or any of the parent directories): .git`

What I suggest is that prettier ignore any file in a `.git` directory.

**Comment by azz** [comment]
What about other version control systems such as mercurial (`.hg/`) and subversion (`.svn/`)? Maybe ignore dot-directories and dotfiles by default?

Alternatively we could remove the default parser and rely entirely on extension or `--parser` option, this way `.git/HEAD` will not match a parser and won't be modified, this might solve a more general case.

But you probably shouldn't be running prettier on a glob without extension(s).

**Comment by lydell** [comment]
@testerez Thanks for the steps to reproduce!

For reference, here are the modifications in a fresh git repo:

.git/HEAD
```diff
-ref: refs/heads/master
+ref: refs / heads / master;
```

.git/config:
```diff
-[core]
-	repositoryformatversion = 0
-	filemode = true
-	bare = false
-	logallrefupdates = true
+[core];
+repositoryformatversion = 0;
+filemode = true;
+bare = false;
+logallrefupdates = true;
```

**Comment by testerez** [comment]
@azz I tried this command after reading this in the doc: "Prettier automatically infers the parser from the input file path"
You're probably right that I shouldn't have run it that way but I think that it can be pretty common mistake.

Thanks for the diff @lydell

**Comment by lydell** [comment]
> Prettier automatically infers the parser from the input file path

Yeah, that could be considered a bit misleading since Prettier falls back to treating everything as JS if it can't infer the type of the file.

**Comment by ikatyang** [comment]
This should be fixed by #4528.
